### PR TITLE
fix ModelContainer.get_crds_parameters

### PR DESCRIPTION
**Pull request by braingram**
`ModelContainer.get_crds_parameters` fails if called on a `ModelContainer` not read from an association.

Examples of failures can be found in the errors in a test run:
https://github.com/spacetelescope/stpipe/actions/runs/8008830118/job/21876144742
where an attempt was made to update `stpipe` to use `ModelContainer.get_crds_parameters`.

This PR updates `ModelContainer.get_crds_parameters` to not assume that an association was used to construct the `ModelContainer`.

To demonstrate the change this will allow a draft PR was opened in `stpipe` that switches it to use `ModelContainer.get_crds_parameters`: https://github.com/spacetelescope/stpipe/pull/144

Regtests show 1 error: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1230/
`jwst.associations.tests.test_level3_duplicate.[stable-deps] test_duplicate_names` failed due to a missing log message. This commonly fails during regression runs likely due to `stpipe` handling of loggers. Note that this is not a regtest and passed in the CI runs.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] ~updated or added relevant tests~ code is at the moment unused (see above)
- [x] ~updated relevant documentation~ no documentation changes were needed
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/datamodels/container.py
Comment (hbushouse) [inline jwst/datamodels/container.py:535]: The old code specifically looked for and opened the first exposure that has type "SCIENCE" (sometimes BACKGROUND or TARGET_ACQ exposures could be encountered first in an association), while this update is hardwired to always just open the first model in the input list. Is that going to work OK?

Comment (braingram) [reply jwst/datamodels/container.py:535]: Thanks for taking a look!

You're spot on that this changes the behavior of this function. However this function is unused. It's not covered by any tests and the part of stpipe that likely should use this function instead grabs the first model (regardless of exposure type):
For pipeline: https://github.com/spacetelescope/stpipe/blob/824863ed2f5c833e0ccc9a3ab7c1c52cce593572/src/stpipe/pipeline.py#L187
For step: https://github.com/spacetelescope/stpipe/blob/824863ed2f5c833e0ccc9a3ab7c1c52cce593572/src/stpipe/step.py#L847

If, as you noted, only the "SCIENCE" exposures should be used than this is a bug in either ModelContainer or stpipe (as it's not currently using that selection criteria). `_open_first_science_exposure` also fails for any `ModelContainer` not initialized from an association (like one made from a list of models). I saw many of these failures when attempting to switch stpipe to used `ModelContainer.get_crds_parameters`.

This PR will allow https://github.com/spacetelescope/stpipe/pull/144 which changes stpipe to used `ModelContainer.get_crds_parameters` (which should no longer produce errors due to this change). This would then allow `jwst` and `romancal` to implement their own logic for selecting the parameters based on the available models (such as selecting the first "SCIENCE" or perhaps just the first model if no exposure type is known).

Comment (jdavies-st) [reply jwst/datamodels/container.py:535]: Maybe it should try to use the first science exposure, but if an exception is raised, handle it, and fall back on opening the first exposure?

```python
try:
    _open_first_science_exposure()
except AttributeError:
   _open_first_exposure()
```

Comment (nden) [reply jwst/datamodels/container.py:535]: I agree there's a chance it may get the wrong CRDS parameters if not a science exposure is used.
Since `_open_first_science_exposure` does not actually open all files and just parses the asn file, is it too much overhead to keep using it? There's a good chance a science exposure will be towards the beginning of the file and if not, oh well ...

Comment (nden) [reply jwst/datamodels/container.py:535]: I think I missed the fact that you are trying to solve a problem when the ModelContainer was not generated from an asn. :-(

Comment (braingram) [reply jwst/datamodels/container.py:535]: Thanks all for your input on this.

As far as I can tell `ModelContainer.get_crds_parameters` is currently unused and the pipeline will use the first model in the association (regardless of exposure type).

Looking for uses of `get_crds_parameters` it isn't used by unit tests:
https://app.codecov.io/gh/spacetelescope/jwst/blob/master/jwst%2Fdatamodels%2Fcontainer.py#L526
or in regtest runs (as seen in the "nightly" coverage):
https://app.codecov.io/gh/spacetelescope/jwst/blob/master/jwst%2Fdatamodels%2Fcontainer.py?flags%5B0%5D=nightly#L526

For `pars-` file selection, stpipe uses the first model (regardless of type) by indexing into the container (and not uses `get_crds_parameters):
- in step: https://github.com/spacetelescope/stpipe/blob/b928337fc3579c025d59073bdf054f133f49a728/src/stpipe/step.py#L862-L866
- in pipeline: https://github.com/spacetelescope/stpipe/blob/b928337fc3579c025d59073bdf054f133f49a728/src/stpipe/pipeline.py#L185-L188

`Step.get_reference_file` can theoretically call `get_crds_parameters` on a `ModelContainer`. However I have yet to find a use in the jwst code and for example `ResampleStep` explicitly uses the index 0 model:
https://github.com/spacetelescope/jwst/blob/5ca492c2ad918f9a4e9b9a53eb21c6af95c5d9d3/jwst/resample/resample_step.py#L93

I've opened a test PR that intentionally breaks `get_crds_parameters` (and `_open_first_science_exposure`):
https://github.com/spacetelescope/jwst/pull/8358
it has so far passed all unit tests and I expect will pass regression tests given the above coverage report (I'll try to start a run when jenkins is free).

It sounds like there is an issue with the current behavior (without this PR) where the pipeline should be using the first "SCIENCE" exposure. The changes in this PR aimed to not modify the reference file selection behavior and instead provide a consistent (and working) interface for stpipe to use.

Should this PR be on hold until the issues with the pipeline reference file selection are addressed? If so, I'm happy to close/convert it to draft.

Comment (hbushouse) [reply jwst/datamodels/container.py:535]: I think you've convinced me that blind usage of the first model/exposure is at least OK for now, because that's what is actually being done now and everything seems to work OK. I wonder if @stscieisenhamer could tell us whether there's logic in the asn generator that *always* puts a science exposure first in an ASN file. I've never seen an ASN file that does *not* have a science exposure at the top, but users could potentially play around with asn files on their own and shuffle the ordering. Or someone could manually build a `ModelContainer` that has a non-science exposure in the first slot. But IMHO that's in the realm of "buyer be ware" and the code should be OK as is.

**Review by hbushouse — APPROVED** [review]
Looks good to me.

**Comment by hbushouse** [comment]
@nden @jdavies-st Any final comments or suggestions before we merge this?

**Comment by nden** [comment]
I still think this is probably a bug in the current pipeline but we don't see it because the Level 3 pipeline opens an association by passing `asn_exptypes = ['science']`. It's also probably true that the "interesting" metadata that `get_crds_parameters` provides are in most cases identical between different association exptype. 
@hbushouse Is there a use case where a non-science exposure in an association file will have different metadata from the science exposure? And by "metadata" the only interest is in those keywords used as selectors in CRDS.
If the input to ModelContainer is not an association, opening the first model should be fine.

**Comment by tapastro** [comment]
I would imagine a level 2 association with target acquisitions would qualify - perhaps also coronagraphic associations? Not sure if the background, science and PSF exposures have differing exp_types, but they certainly might.

**Comment by braingram** [comment]
After some discussion with @nden I'd like to delay this change until after the release (and will make this PR draft). The main motivation is to address the above comments and change`get_crds_parameters` to use the first science exposure type. Some updates that may be added are:

- modifying `get_crds_parameters` to use exposure type when loaded from an association backed and hopefully from a model list
- modifying steps that explicitly use the first index model to instead select by exposure type
- tests for the above behavior including associations where the first listed model is not science and has different metadata

The above will expand the scope of this PR and I'll bring it back out of draft and re-request reviews when it's ready.

Thanks all for the discussion and comments.

@nden and @tapastro having example associations with models with varying metadata would be excellent! Please feel free to send me links/files.

**Pull request closed without merging**
